plugins/kinetic-dp/fu-kinetic-dp-puma-device.c

```diff
@@ -0,0 +1,563 @@
+		fu_device_sleep(FU_DEVICE(self), 18);
+		if (!fu_device_retry_full(
+			FU_DEVICE(self),
+			fu_kinetic_dp_puma_device_wait_dpcd_cmd_status_cb,
+			150,
+			POLL_INTERVAL_MS,
+			GUINT_TO_POINTER(FU_KINETIC_DP_PUMA_MODE_FLASH_INFO_READY),
+			error)) {
+			g_prefix_error(error, "timeout waiting for MODE_FLASH_INFO_READY: ");
+			return FALSE;
+		}
+
+		/* get flash info */
+		if (!fu_dpaux_device_read(FU_DPAUX_DEVICE(self),
+					  PUMA_DPCD_DATA_ADDR,
+					  flashinfo,
+					  sizeof(flashinfo),
+					  FU_KINETIC_DP_DEVICE_TIMEOUT,
+					  error)) {
+			g_prefix_error(error, "failed to read Flash Info: ");
+			return FALSE;
+		}
+		st =
+		    fu_struct_kinetic_dp_flash_info_parse(flashinfo, sizeof(flashinfo), 0x0, error);
+		self->flash_id = fu_struct_kinetic_dp_flash_info_get_id(st);
+		self->flash_size = fu_struct_kinetic_dp_flash_info_get_size(st);
+		self->read_flash_prog_time = fu_struct_kinetic_dp_flash_info_get_erase_time(st);
+
+		/* save flash info need to do memcopy copy here */
+		if (self->flash_size == 0) {
+			if (self->flash_id > 0) {
+				g_set_error_literal(error,
+						    G_IO_ERROR,
+						    G_IO_ERROR_NOT_SUPPORTED,
+						    "SPI flash not supported");
+				return FALSE;
+			}
+			g_set_error_literal(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_NOT_SUPPORTED,
+					    "SPI flash not connected");
+			return FALSE;
+		}
+		if (self->flash_size >= 0x400)
+			fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_DUAL_IMAGE);
+	}
+
+	/* checking for flash erase done */
+	g_debug("waiting for flash erasing...");
+	if (self->read_flash_prog_time)
+		fu_device_sleep(FU_DEVICE(self), self->read_flash_prog_time);
+	else
+		fu_device_sleep(FU_DEVICE(self), FU_KINETIC_DP_PUMA_REQUEST_FLASH_ERASE_TIME);
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_kinetic_dp_puma_device_wait_dpcd_sink_mode_cb,
+				  150,
+				  POLL_INTERVAL_MS,
+				  GUINT_TO_POINTER(FU_KINETIC_DP_PUMA_REQUEST_FW_UPDATE_READY),
+				  error)) {
+		g_prefix_error(error, "timeout waiting for REQUEST_FW_UPDATE_READY: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_kinetic_dp_puma_device_setup(FuDevice *device, GError **error)
+{
+	FuKineticDpPumaDevice *self = FU_KINETIC_DP_PUMA_DEVICE(device);
+	guint8 dpcd_buf[3] = {0};
+	g_autofree gchar *version = NULL;
+
+	/* FuKineticDpDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_kinetic_dp_puma_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* read major and minor version */
+	if (!fu_dpaux_device_read(FU_DPAUX_DEVICE(self),
+				  DPCD_ADDR_BRANCH_FW_MAJ_REV,
+				  dpcd_buf,
+				  2,
+				  FU_KINETIC_DP_DEVICE_TIMEOUT,
+				  error))
+		return FALSE;
+
+	/* read sub */
+	if (!fu_dpaux_device_read(FU_DPAUX_DEVICE(self),
+				  DPCD_ADDR_BRANCH_FW_SUB,
+				  dpcd_buf + 2,
+				  1,
+				  FU_KINETIC_DP_DEVICE_TIMEOUT,
+				  error))
+		return FALSE;
+	version = g_strdup_printf("%1d.%03d.%02d", dpcd_buf[0], dpcd_buf[1], dpcd_buf[2]);
```

Comment:
this is a really weird format, can we not align it to something more standard?

Reply:
I believe it's what they use already for Windows. It's technically a triplet, but I agree it's weird.